File: src/main/java/seremis/geninfusion/api/soul/util/ModelPart.java
package seremis.geninfusion.api.soul.util;

import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;
import net.minecraft.client.model.*;
import net.minecraft.entity.EntityLiving;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.nbt.NBTTagList;
import net.minecraft.util.Vec3;
import net.minecraftforge.common.util.Constants;
import seremis.geninfusion.api.soul.SoulHelper;
import seremis.geninfusion.helper.GIReflectionHelper;
import seremis.geninfusion.util.INBTTagable;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class ModelPart extends ModelRenderer implements INBTTagable {

    public float initialRotationPointX, initialRotationPointY, initialRotationPointZ;
    public float initialRotateAngleX, initialRotateAngleY, initialRotateAngleZ;

    public ModelPart(String boxName) {
        super(SoulHelper.entityModel(), boxName);
    }

    public ModelPart() {
        super(SoulHelper.entityModel());
    }

    public ModelPart(int textureOffsetX, int textureOffsetY) {
        super(SoulHelper.entityModel(), textureOffsetX, textureOffsetY);
    }

    public ModelPart(NBTTagCompound compound) {
        this(compound.getString("boxName"));
        readFromNBT(compound);
    }

    private boolean firstTick = true;

    @Override
    @SideOnly(Side.CLIENT)
    public void render(float scale) {
        if(firstTick) {
            initialRotationPointX = rotationPointX;
            initialRotationPointY = rotationPointY;
            initialRotationPointZ = rotationPointZ;
            initialRotateAngleX = rotateAngleX;
            initialRotateAngleY = rotateAngleY;
            initialRotateAngleZ = rotateAngleZ;
            firstTick = false;
        }
        super.render(scale);
    }


    public Vec3 getInitialRotationPoints() {
        return Vec3.createVectorHelper(initialRotationPointX, initialRotationPointY, initialRotationPointZ);
    }

    public Vec3 getInitialRotateAngles() {
        return Vec3.createVectorHelper(initialRotateAngleX, initialRotateAngleY, initialRotateAngleZ);
    }

    public int getTextureOffsetX() {
        return (Integer) GIReflectionHelper.getField(this, "textureOffsetX");
    }

    public int getTextureOffsetY() {
        return (Integer) GIReflectionHelper.getField(this, "textureOffsetY");
    }

    public List<ModelBox> getBoxList() {
        return cubeList;
    }

    public TexturedQuad[] getBoxQuads(ModelBox box) {
        return (TexturedQuad[]) GIReflectionHelper.getField(box, "quadList");
    }

    @Override
    public NBTTagCompound writeToNBT(NBTTagCompound compound) {
        if(boxName != null && !boxName.equals(""))
            compound.setString("boxName", boxName);
        compound.setFloat("textureWidth", textureWidth);
        compound.setFloat("textureHeight", textureHeight);
        compound.setFloat("rotationPointX", rotationPointX);
        compound.setFloat("rotationPointY", rotationPointY);
        compound.setFloat("rotationPointZ", rotationPointZ);
        compound.setFloat("rotateAngleX", rotateAngleX);
        compound.setFloat("rotateAngleY", rotateAngleY);
        compound.setFloat("rotateAngleZ", rotateAngleZ);
        compound.setBoolean("mirror", mirror);
        compound.setBoolean("showModel", showModel);
        compound.setBoolean("isHidden", isHidden);
        compound.setFloat("offsetX", offsetX);
        compound.setFloat("offsetY", offsetY);
        compound.setFloat("offsetZ", offsetZ);

        NBTTagList childList = new NBTTagList();

        if(childModels != null) {
            for(Object childModel : childModels) {
                ModelPart modelPart = ModelPart.modelRendererToModelPart((ModelRenderer) childModel);
                NBTTagCompound compound1 = new NBTTagCompound();

                modelPart.writeToNBT(compound1);

                childList.appendTag(compound1);
            }
        }
        compound.setTag("childModels", childList);

        NBTTagList boxList = new NBTTagList();

        if(cubeList != null) {
            for(Object cube : cubeList) {
                ModelBox box = (ModelBox) cube;
                NBTTagCompound boxCompound = new NBTTagCompound();

                if(box.field_78247_g != null && !box.field_78247_g.equals(""))
                    compound.setString("boxName", box.field_78247_g);
                boxCompound.setFloat("originX", box.posX1);
                boxCompound.setFloat("originY", box.posY1);
                boxCompound.setFloat("originZ", box.posZ1);
                boxCompound.setInteger("sizeX", (int) (box.posX2 - box.posX1));
                boxCompound.setInteger("sizeY", (int) (box.posY2 - box.posY1));
                boxCompound.setInteger("sizeZ", (int) (box.posZ2 - box.posZ1));
                TexturedQuad[] quads = (TexturedQuad[]) GIReflectionHelper.getField(box, "quadList");

                NBTTagList quadList = new NBTTagList();

                for(TexturedQuad quad : quads) {
                    NBTTagList vertexList = new NBTTagList();

                    PositionTextureVertex[] vertices = quad.vertexPositions;

                    for(PositionTextureVertex vertex : vertices) {
                        NBTTagCompound vertexCompound = new NBTTagCompound();

                        vertexCompound.setDouble("vectorX", vertex.vector3D.xCoord);
                        vertexCompound.setDouble("vectorY", vertex.vector3D.yCoord);
                        vertexCompound.setDouble("vectorZ", vertex.vector3D.zCoord);
                        vertexCompound.setFloat("texturePositionX", vertex.texturePositionX);
                        vertexCompound.setFloat("texturePositionY", vertex.texturePositionY);

                        vertexList.appendTag(vertexCompound);
                    }
                    NBTTagCompound vertexListCompound = new NBTTagCompound();
                    vertexListCompound.setTag("vertexList", vertexList);

                    quadList.appendTag(vertexListCompound);
                }
                boxCompound.setTag("quadList", quadList);


                NBTTagList vertexList = new NBTTagList();
                PositionTextureVertex[] vertices = (PositionTextureVertex[]) GIReflectionHelper.getField(box, "vertexPositions");

                for(PositionTextureVertex vertex : vertices) {
                    NBTTagCompound vertexCompound = new NBTTagCompound();

                    vertexCompound.setDouble("vectorX", vertex.vector3D.xCoord);
                    vertexCompound.setDouble("vectorY", vertex.vector3D.yCoord);
                    vertexCompound.setDouble("vectorZ", vertex.vector3D.zCoord);
                    vertexCompound.setFloat("texturePositionX", vertex.texturePositionX);
                    vertexCompound.setFloat("texturePositionY", vertex.texturePositionY);

                    vertexList.appendTag(vertexCompound);
                }

                boxCompound.setTag("vertexList", vertexList);

                boxList.appendTag(boxCompound);
            }
        }
        compound.setTag("cubeList", boxList);
        return compound;
    }

    @Override
    public NBTTagCompound readFromNBT(NBTTagCompound compound) {
        if(compound.hasKey("boxName"))
            GIReflectionHelper.setField(this, "boxName", compound.getString("boxName"));

        textureWidth = compound.getFloat("textureWidth");
        textureHeight = compound.getFloat("textureHeight");
        rotationPointX = compound.getFloat("rotationPointX");
        rotationPointY = compound.getFloat("rotationPointY");
        rotationPointZ = compound.getFloat("rotationPointZ");
        rotateAngleX = compound.getFloat("rotateAngleX");
        rotateAngleY = compound.getFloat("rotateAngleY");
        rotateAngleZ = compound.getFloat("rotateAngleZ");
        mirror = compound.getBoolean("mirror");
        showModel = compound.getBoolean("showModel");
        isHidden = compound.getBoolean("isHidden");
        offsetX = compound.getFloat("offsetX");
        offsetY = compound.getFloat("offsetY");
        offsetZ = compound.getFloat("offsetZ");

        NBTTagList childList = (NBTTagList) compound.getTag("childModels");
        for(int i = 0; i < childList.tagCount(); i++) {
            ModelPart childPart = new ModelPart(childList.getCompoundTagAt(i));
            addChild(childPart);
        }

        NBTTagList boxList = (NBTTagList) compound.getTag("cubeList");
        for(int i = 0; i < boxList.tagCount(); i++) {
            NBTTagCompound boxCompound = boxList.getCompoundTagAt(i);
            ModelBox box = new ModelBox(this, (Integer) GIReflectionHelper.getField(this, "textureOffsetX"), (Integer) GIReflectionHelper.getField(this, "textureOffsetY"), boxCompound.getFloat("originX"), boxCompound.getFloat("originY"), boxCompound.getFloat("originZ"), boxCompound.getInteger("sizeX"), boxCompound.getInteger("sizeY"), boxCompound.getInteger("sizeZ"), 0.0F);

            NBTTagList quadList = (NBTTagList) boxCompound.getTag("quadList");

            TexturedQuad[] quads = new TexturedQuad[quadList.tagCount()];

            for(int j = 0; j < quadList.tagCount(); j++) {
                NBTTagCompound vertexListCompound = quadList.getCompoundTagAt(j);
                NBTTagList vertexList = vertexListCompound.getTagList("vertexList", Constants.NBT.TAG_COMPOUND);

                PositionTextureVertex[] vertices = new PositionTextureVertex[vertexList.tagCount()];

                for(int k = 0; k < vertexList.tagCount(); k++) {
                    NBTTagCompound vertexCompound = vertexList.getCompoundTagAt(k);

                    Vec3 vector = Vec3.createVectorHelper(vertexCompound.getDouble("vectorX"), vertexCompound.getDouble("vectorY"), vertexCompound.getDouble("vectorZ"));
                    vertices[k] = new PositionTextureVertex(vector, vertexCompound.getFloat("texturePositionX"), vertexCompound.getFloat("texturePositionY"));
                }

                quads[j] = new TexturedQuad(vertices);
            }

            GIReflectionHelper.setField(box, "quadList", quads);


            NBTTagList vertexList = boxCompound.getTagList("vertexList", Constants.NBT.TAG_COMPOUND);

            PositionTextureVertex[] vertices = new PositionTextureVertex[vertexList.tagCount()];

            for(int k = 0; k < vertexList.tagCount(); k++) {
                NBTTagCompound vertexCompound = vertexList.getCompoundTagAt(k);

                Vec3 vector = Vec3.createVectorHelper(vertexCompound.getDouble("vectorX"), vertexCompound.getDouble("vectorY"), vertexCompound.getDouble("vectorZ"));
                vertices[k] = new PositionTextureVertex(vector, vertexCompound.getFloat("texturePositionX"), vertexCompound.getFloat("texturePositionY"));
            }

            GIReflectionHelper.setField(box, "vertexPositions", vertices);

            cubeList.add(box);
        }
        return compound;
    }

    private Random rand = new Random();

    public ModelPart mutate() {
        switch(rand.nextInt(2)) {
            case 0:
                rotationPointX = (float) (rotationPointX * ((rand.nextFloat() * 2 * 0.1) + 0.9));
                rotationPointY = (float) (rotationPointY * ((rand.nextFloat() * 2 * 0.1) + 0.9));
                rotationPointZ = (float) (rotationPointZ * ((rand.nextFloat() * 2 * 0.1) + 0.9));
                break;
            case 1:
                rotateAngleX = (float) (rotateAngleX * ((rand.nextFloat() * 2 * 0.1) + 0.9));
                rotateAngleY = (float) (rotateAngleY * ((rand.nextFloat() * 2 * 0.1) + 0.9));
                rotateAngleZ = (float) (rotateAngleZ * ((rand.nextFloat() * 2 * 0.1) + 0.9));
                break;
        }
        return this;
    }

    public static ModelPart[] getModelPartsFromModel(ModelBase model, EntityLiving entity) {
        ArrayList<ModelPart> parts = new ArrayList<ModelPart>();

        //With this skeletons animate without zombie arms without having bows in their hands
        if(!(model instanceof ModelSkeleton)) {
            model.setRotationAngles(0.0F, 0.0F, 0.0F, 0.0F, 0.0F, 0.0F, entity);
        }

        Field[] fields = GIReflectionHelper.getFields(model);

        for(Field field : fields) {
            if(field.getType().equals(ModelRenderer.class) && !field.getName().equals("bipedCloak") && !field.getName().equals("bipedEars")) {
                ModelRenderer renderer = (ModelRenderer) GIReflectionHelper.getField(model, field.getName());
                parts.add(modelRendererToModelPart(renderer));
            }
        }

        return parts.toArray(new ModelPart[parts.size()]);
    }

    public static ModelPart modelRendererToModelPart(ModelRenderer model) {
        ModelPart modelPart = new ModelPart(model.boxName);
        modelPart.childModels = model.childModels;
        modelPart.cubeList = model.cubeList;
        modelPart.isHidden = model.isHidden;
        modelPart.showModel = model.showModel;
        modelPart.mirror = model.mirror;
        modelPart.rotationPointX = model.rotationPointX;
        modelPart.rotationPointY = model.rotationPointY;
        modelPart.rotationPointZ = model.rotationPointZ;
        modelPart.rotateAngleX = model.rotateAngleX;
        modelPart.rotateAngleY = model.rotateAngleY;
        modelPart.rotateAngleZ = model.rotateAngleZ;
        modelPart.textureWidth = model.textureWidth;
        modelPart.textureHeight = model.textureHeight;
        modelPart.offsetX = model.offsetX;
        modelPart.offsetY = model.offsetY;
        modelPart.offsetZ = model.offsetZ;
        modelPart.setTextureOffset((Integer) GIReflectionHelper.getField(model, "textureOffsetX"), (Integer) GIReflectionHelper.getField(model, "textureOffsetY"));
        return modelPart;
    }

    public static ModelPart fromNBT(NBTTagCompound compound) {
        ModelPart part = new ModelPart();
        part.readFromNBT(compound);
        return part;
    }

    @Override
    public boolean equals(Object obj) {
        if(obj instanceof ModelPart) {
            ModelPart part = (ModelPart) obj;
            NBTTagCompound compound1 = new NBTTagCompound();
            NBTTagCompound compound2 = new NBTTagCompound();
            this.writeToNBT(compound1);
            part.writeToNBT(compound2);
            return compound1.equals(compound2);
        }
        return false;
    }
}

File: src/main/scala/seremis/geninfusion/api/util/render/model/ModelPart.scala
package seremis.geninfusion.api.util.render.model

import java.util
import java.util.Random

import cpw.mods.fml.relauncher.{Side, SideOnly}
import net.minecraft.client.model._
import net.minecraft.client.renderer.GLAllocation
import net.minecraft.entity.{Entity, EntityLiving}
import net.minecraft.nbt.{NBTTagCompound, NBTTagList}
import net.minecraft.util.Vec3
import net.minecraftforge.common.util.Constants
import seremis.geninfusion.api.soul.SoulHelper
import seremis.geninfusion.api.soul.lib.VariableLib
import seremis.geninfusion.helper.GIReflectionHelper
import seremis.geninfusion.util.{GIModelBox, INBTTagable}

import scala.collection.JavaConversions._
import scala.collection.mutable.ListBuffer

object ModelPart {

    private def getModelPartsFromModel(model: ModelBase, entity: EntityLiving, callRotationAngles: Boolean): Array[ModelPart] = {
        val parts: ListBuffer[ModelPart] = ListBuffer()

        model.swingProgress = 0

        if (callRotationAngles && !model.isInstanceOf[ModelSkeleton]) {
            model.setRotationAngles(0.0F, 0.0F, 0.0F, 0.0F, 0.0F, 0.0F, null.asInstanceOf[Entity])
        }

        val fields = GIReflectionHelper.getFields(model)

        for (field <- fields if (field.getType == classOf[ModelRenderer] || field.getType == classOf[ModelPart]) && field.getName != VariableLib.ModelBipedCloak && field.getName != VariableLib.ModelBipedEars && field.getName != VariableLib.ModelBipedHeadwear) {
           // parts.add(modelRendererToModelPart(GIReflectionHelper.getField(model, field.getName).asInstanceOf[ModelRenderer]))
        }
        parts.to[Array]
    }

    def rendererToPart(model: ModelRenderer, modelPartType: String): ModelPart = {
        val modelPart = new ModelPart(GIReflectionHelper.getField(model, VariableLib.ModelRendererBaseModel).asInstanceOf[ModelBase], model.boxName, modelPartType)
        modelPart.childModels = model.childModels
        modelPart.cubeList = model.cubeList
        modelPart.isHidden = model.isHidden
        modelPart.showModel = model.showModel
        modelPart.mirror = model.mirror
        modelPart.rotationPointX = model.rotationPointX
        modelPart.rotationPointY = model.rotationPointY
        modelPart.rotationPointZ = model.rotationPointZ
        modelPart.rotateAngleX = model.rotateAngleX
        modelPart.rotateAngleY = model.rotateAngleY
        modelPart.rotateAngleZ = model.rotateAngleZ
        modelPart.textureWidth = model.textureWidth
        modelPart.textureHeight = model.textureHeight
        modelPart.offsetX = model.offsetX
        modelPart.offsetY = model.offsetY
        modelPart.offsetZ = model.offsetZ
        modelPart.setTextureOffset(GIReflectionHelper.getField(model, VariableLib.ModelRendererTextureOffsetX).asInstanceOf[Int], GIReflectionHelper.getField(model, VariableLib.ModelRendererTextureOffsetY).asInstanceOf[Int])
        modelPart
    }

    def fromNBT(compound: NBTTagCompound): ModelPart = {
        val part = new ModelPart(null.asInstanceOf[String])
        part.readFromNBT(compound)
        part
    }
}

class ModelPart(model: ModelBase, boxName: String, var modelPartTypeName: String) extends ModelRenderer(model, boxName) with INBTTagable {

    var modelPartType = if(modelPartTypeName != null) SoulHelper.modelPartTypeRegistry.getModelPartType(modelPartTypeName).get else null

    var initialRotationPointX: Float = 0.0F
    var initialRotationPointY: Float = 0.0F
    var initialRotationPointZ: Float = 0.0F

    var initialRotateAngleX: Float = 0.0F
    var initialRotateAngleY: Float = 0.0F
    var initialRotateAngleZ: Float = 0.0F

    def this(modelPartTypeName: String) {
        this(SoulHelper.entityModel, "", modelPartTypeName)
    }

    def this(boxName: String, modelPartTypeName: String) {
        this(SoulHelper.entityModel, boxName, modelPartTypeName)
    }

    def this(model: ModelBase, modelPartTypeName: String) {
        this(model, "", modelPartTypeName)
    }

    def this(compound: NBTTagCompound) {
        this(null, compound.getString("boxName"), null)
        readFromNBT(compound)
    }

    private var firstTick: Boolean = true

    @SideOnly(Side.CLIENT)
    override def render(scale: Float) {
        if (firstTick) {
            initialRotationPointX = rotationPointX
            initialRotationPointY = rotationPointY
            initialRotationPointZ = rotationPointZ
            initialRotateAngleX = rotateAngleX
            initialRotateAngleY = rotateAngleY
            initialRotateAngleZ = rotateAngleZ
            firstTick = false
        }
        super.render(scale)
    }

    def getInitialRotationPoints: Vec3 = {
        Vec3.createVectorHelper(initialRotationPointX, initialRotationPointY, initialRotationPointZ)
    }

    def getInitialRotateAngles: Vec3 = {
        Vec3.createVectorHelper(initialRotateAngleX, initialRotateAngleY, initialRotateAngleZ)
    }

    def getTextureOffsetX: Int = {
        GIReflectionHelper.getField(this, VariableLib.ModelRendererTextureOffsetX).asInstanceOf[java.lang.Integer]
    }

    def getTextureOffsetY: Int = {
        GIReflectionHelper.getField(this, VariableLib.ModelRendererTextureOffsetY).asInstanceOf[java.lang.Integer]
    }

    def getBoxList: Array[ModelBox] = cubeList.asInstanceOf[util.ArrayList[ModelBox]].to[Array]

    def getBoxQuads(box: ModelBox): Array[TexturedQuad] = {
        GIReflectionHelper.getField(box, VariableLib.ModelBoxQuadList).asInstanceOf[Array[TexturedQuad]]
    }

    def addGIBox(x: Float, y: Float, z: Float, sizeX: Float, sizeY: Float, sizeZ: Float) {
        cubeList.asInstanceOf[util.List[ModelBox]].add(new GIModelBox(this, getTextureOffsetX, getTextureOffsetY, x, y, z, sizeX, sizeY, sizeZ, 0.0F))
    }

    override def writeToNBT(compound: NBTTagCompound): NBTTagCompound = {
        if (boxName != null && boxName != "") compound.setString("boxName", boxName)
        compound.setFloat("textureWidth", textureWidth)
        compound.setFloat("textureHeight", textureHeight)
        compound.setFloat("rotationPointX", rotationPointX)
        compound.setFloat("rotationPointY", rotationPointY)
        compound.setFloat("rotationPointZ", rotationPointZ)
        compound.setFloat("rotateAngleX", rotateAngleX)
        compound.setFloat("rotateAngleY", rotateAngleY)
        compound.setFloat("rotateAngleZ", rotateAngleZ)
        compound.setBoolean("mirror", mirror)
        compound.setBoolean("showModel", showModel)
        compound.setBoolean("isHidden", isHidden)
        compound.setFloat("offsetX", offsetX)
        compound.setFloat("offsetY", offsetY)
        compound.setFloat("offsetZ", offsetZ)
        val childList = new NBTTagList()
        if (childModels != null) {
            for (childModel <- childModels) {
                val modelPart = childModel.asInstanceOf[ModelPart]
                val compound1 = new NBTTagCompound()
                modelPart.writeToNBT(compound1)
                childList.appendTag(compound1)
            }
        }
        compound.setTag("childModels", childList)
        val boxList = new NBTTagList()
        if (cubeList != null) {
            for (cube <- cubeList) {
                val box = cube.asInstanceOf[ModelBox]
                val boxCompound = new NBTTagCompound()
                if (box.field_78247_g != null && box.field_78247_g != "") compound.setString("boxName", box.field_78247_g)
                boxCompound.setFloat("originX", box.posX1)
                boxCompound.setFloat("originY", box.posY1)
                boxCompound.setFloat("originZ", box.posZ1)
                boxCompound.setInteger("sizeX", (box.posX2 - box.posX1).toInt)
                boxCompound.setInteger("sizeY", (box.posY2 - box.posY1).toInt)
                boxCompound.setInteger("sizeZ", (box.posZ2 - box.posZ1).toInt)
                val quads = GIReflectionHelper.getField(box, VariableLib.ModelBoxQuadList).asInstanceOf[Array[TexturedQuad]]
                val quadList = new NBTTagList()
                for (quad <- quads) {
                    val vertexList = new NBTTagList()
                    val vertices = quad.vertexPositions
                    for (vertex <- vertices) {
                        val vertexCompound = new NBTTagCompound()
                        vertexCompound.setDouble("vectorX", vertex.vector3D.xCoord)
                        vertexCompound.setDouble("vectorY", vertex.vector3D.yCoord)
                        vertexCompound.setDouble("vectorZ", vertex.vector3D.zCoord)
                        vertexCompound.setFloat("texturePositionX", vertex.texturePositionX)
                        vertexCompound.setFloat("texturePositionY", vertex.texturePositionY)
                        vertexList.appendTag(vertexCompound)
                    }
                    val vertexListCompound = new NBTTagCompound()
                    vertexListCompound.setTag("vertexList", vertexList)
                    quadList.appendTag(vertexListCompound)
                }
                boxCompound.setTag("quadList", quadList)
                val vertexList = new NBTTagList()
                val vertices = GIReflectionHelper.getField(box, VariableLib.ModelBoxVertexPositions).asInstanceOf[Array[PositionTextureVertex]]
                for (vertex <- vertices) {
                    val vertexCompound = new NBTTagCompound()
                    vertexCompound.setDouble("vectorX", vertex.vector3D.xCoord)
                    vertexCompound.setDouble("vectorY", vertex.vector3D.yCoord)
                    vertexCompound.setDouble("vectorZ", vertex.vector3D.zCoord)
                    vertexCompound.setFloat("texturePositionX", vertex.texturePositionX)
                    vertexCompound.setFloat("texturePositionY", vertex.texturePositionY)
                    vertexList.appendTag(vertexCompound)
                }
                boxCompound.setTag("vertexList", vertexList)
                boxList.appendTag(boxCompound)
            }
        }
        compound.setTag("cubeList", boxList)

        compound.setString("modelPartType", modelPartTypeName)

        compound
    }

    override def readFromNBT(compound: NBTTagCompound): NBTTagCompound = {
        if (compound.hasKey("boxName")) GIReflectionHelper.setField(this, VariableLib.ModelRendererBoxName,
            compound.getString("boxName"))
        textureWidth = compound.getFloat("textureWidth")
        textureHeight = compound.getFloat("textureHeight")
        rotationPointX = compound.getFloat("rotationPointX")
        rotationPointY = compound.getFloat("rotationPointY")
        rotationPointZ = compound.getFloat("rotationPointZ")
        rotateAngleX = compound.getFloat("rotateAngleX")
        rotateAngleY = compound.getFloat("rotateAngleY")
        rotateAngleZ = compound.getFloat("rotateAngleZ")
        mirror = compound.getBoolean("mirror")
        showModel = compound.getBoolean("showModel")
        isHidden = compound.getBoolean("isHidden")
        offsetX = compound.getFloat("offsetX")
        offsetY = compound.getFloat("offsetY")
        offsetZ = compound.getFloat("offsetZ")

        val childList = compound.getTag("childModels").asInstanceOf[NBTTagList]

        for (i <- 0 until childList.tagCount()) {
            addChild(new ModelPart(childList.getCompoundTagAt(i)))
        }
        val boxList = compound.getTag("cubeList").asInstanceOf[NBTTagList]
        for (i <- 0 until boxList.tagCount()) {
            val boxCompound = boxList.getCompoundTagAt(i)

            val box = new ModelBox(this, GIReflectionHelper.getField(this, VariableLib.ModelRendererTextureOffsetX).asInstanceOf[java.lang.Integer], GIReflectionHelper.getField(this, VariableLib.ModelRendererTextureOffsetY).asInstanceOf[java.lang.Integer], boxCompound.getFloat("originX"), boxCompound.getFloat("originY"), boxCompound.getFloat("originZ"), boxCompound.getInteger("sizeX"), boxCompound.getInteger("sizeY"), boxCompound.getInteger("sizeZ"), 0.0F)

            val quadList = boxCompound.getTag("quadList").asInstanceOf[NBTTagList]
            val quads = Array.ofDim[TexturedQuad](quadList.tagCount())

            for (j <- 0 until quadList.tagCount()) {
                val vertexListCompound = quadList.getCompoundTagAt(j)

                val vertexList = vertexListCompound.getTagList("vertexList", Constants.NBT.TAG_COMPOUND)
                val vertices = Array.ofDim[PositionTextureVertex](vertexList.tagCount())

                for (k <- 0 until vertexList.tagCount()) {
                    val vertexCompound = vertexList.getCompoundTagAt(k)
                    val vector = Vec3.createVectorHelper(vertexCompound.getDouble("vectorX"), vertexCompound.getDouble("vectorY"), vertexCompound.getDouble("vectorZ"))

                    vertices(k) = new PositionTextureVertex(vector, vertexCompound.getFloat("texturePositionX"), vertexCompound.getFloat("texturePositionY"))
                }
                quads(j) = new TexturedQuad(vertices)
            }
            GIReflectionHelper.setField(box, VariableLib.ModelBoxQuadList, quads)

            val vertexList = boxCompound.getTagList("vertexList", Constants.NBT.TAG_COMPOUND)
            val vertices = Array.ofDim[PositionTextureVertex](vertexList.tagCount())

            for (k <- 0 until vertexList.tagCount()) {
                val vertexCompound = vertexList.getCompoundTagAt(k)
                val vector = Vec3.createVectorHelper(vertexCompound.getDouble("vectorX"), vertexCompound.getDouble("vectorY"), vertexCompound.getDouble("vectorZ"))

                vertices(k) = new PositionTextureVertex(vector, vertexCompound.getFloat("texturePositionX"), vertexCompound.getFloat("texturePositionY"))
            }
            GIReflectionHelper.setField(box, VariableLib.ModelBoxVertexPositions, vertices)
            cubeList.asInstanceOf[util.ArrayList[ModelBox]].add(box)
        }

        modelPartTypeName = compound.getString("modelPartType")
        modelPartType = SoulHelper.modelPartTypeRegistry.getModelPartType(modelPartTypeName).get

        compound
    }

    private val rand: Random = new Random()

    def mutate(): ModelPart = {
        rand.nextInt(2) match {
            case 0 =>
                rotationPointX = (rotationPointX * ((rand.nextFloat() * 2 * 0.1) + 0.9)).toFloat
                rotationPointY = (rotationPointY * ((rand.nextFloat() * 2 * 0.1) + 0.9)).toFloat
                rotationPointZ = (rotationPointZ * ((rand.nextFloat() * 2 * 0.1) + 0.9)).toFloat

            case 1 =>
                rotateAngleX = (rotateAngleX * ((rand.nextFloat() * 2 * 0.1) + 0.9)).toFloat
                rotateAngleY = (rotateAngleY * ((rand.nextFloat() * 2 * 0.1) + 0.9)).toFloat
                rotateAngleZ = (rotateAngleZ * ((rand.nextFloat() * 2 * 0.1) + 0.9)).toFloat
        }
        this
    }

    //noinspection ComparingUnrelatedTypes
    override def equals(obj: Any): Boolean = {
        if (obj.isInstanceOf[ModelPart]) {
            val part = obj.asInstanceOf[ModelPart]

            val compound1 = new NBTTagCompound()
            val compound2 = new NBTTagCompound()

            this.writeToNBT(compound1)
            part.writeToNBT(compound2)

            return compound1.equals(compound2)
        }
        false
    }

    def resetDisplayList() {
        if(GIReflectionHelper.getField(this, VariableLib.ModelRendererCompiled).asInstanceOf[Boolean]) {
            GLAllocation.deleteDisplayLists(GIReflectionHelper.getField(this, VariableLib.ModelRendererDisplayList).asInstanceOf[Int])
            GIReflectionHelper.setField(this, VariableLib.ModelRendererCompiled, false)
        }
    }
}